File: src/main/java/com/estonianport/agendaza/model/Usuario.java
package com.estonianport.agendaza.model;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import lombok.Getter;
import lombok.Setter;

@Entity
@Getter
@Setter
public class Usuario implements UserDetails{

	private static final long serialVersionUID = -3874297176481741405L;

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@Column
	private String nombre;

	@Column
	private String apellido;

	@Column
	private String username;

	@Column
	private String password;

	@Column
	private String mail;

	@Column
	private boolean accountNonExpired;

	@Column
	private boolean accountNonLocked;

	@Column
	private boolean credentialsNonExpired;

	@Column
	private boolean enabled;

	@ManyToOne
	@JoinColumn(name = "rol_id")
	private Rol rol;

	@Override
	public Collection<? extends GrantedAuthority> getAuthorities() {
		List<SimpleGrantedAuthority> authorities = new ArrayList<>();
		authorities.add(new SimpleGrantedAuthority(rol.getNombre()));
		return authorities;
	}

}

File: src/main/kotlin/com/estonianport/agendaza/model/Usuario.kt
package com.estonianport.agendaza.model

import com.estonianport.agendaza.dto.ClienteDTO
import com.estonianport.agendaza.dto.UsuarioAbmDTO
import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Index
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import java.time.LocalDate

@Entity
@Table(
    name = "usuario",
    indexes = [
        Index(name = "idx_usuario_username", columnList = "username", unique = true),
        Index(name = "idx_usuario_email_baja", columnList = "email,fecha_baja"),
        Index(name = "idx_usuario_celular_baja", columnList = "celular,fecha_baja")
    ]
)
class Usuario(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long,

    @Column(nullable = false)
    var nombre: String,

    @Column(nullable = false)
    var apellido: String,

    @Column(nullable = false)
    var celular: Long,

    @Column(nullable = false)
    var email: String) {

    @Column(nullable = true)
    var username: String? = null

    @Column(nullable = true)
    var password: String? = null

    @Column
    var fechaNacimiento: LocalDate = LocalDate.now()

    @Column(nullable = false, updatable = false)
    var fechaAlta: LocalDate = LocalDate.now()

    @Column
    var fechaBaja: LocalDate? = null

    @JsonIgnore
    @OneToMany(mappedBy = "usuario", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    var listaCargo: MutableSet<Cargo> = mutableSetOf()

}
